### PR TITLE
introducing azure_rm_resource

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
This is a generic Azure REST module

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
azure_rm_resource
azure_rm_resource_facts

##### ANSIBLE VERSION
```
2.5
```

### CODE REVIEW

File: test/sanity/validate-modules/ignore.txt [hunk 1]
```diff
@@ -198,6 +198,8 @@ lib/ansible/modules/cloud/azure/azure_rm_postgresqlserver.py E326
 lib/ansible/modules/cloud/azure/azure_rm_publicipaddress.py E322
 lib/ansible/modules/cloud/azure/azure_rm_publicipaddress.py E324
 lib/ansible/modules/cloud/azure/azure_rm_publicipaddress_facts.py E324
+lib/ansible/modules/cloud/azure/azure_rm_resource.py E324
+lib/ansible/modules/cloud/azure/azure_rm_resource_facts.py E324
```
Comment (mattclay) [inline test/sanity/validate-modules/ignore.txt]: New ignores should not be added. The issues should be fixed instead. If the errors are due to module_utils or docs fragments, those should be fixed in a separate PR before merging this one.

Comment (zikalino) [reply test/sanity/validate-modules/ignore.txt]: ok, so i will create that separate pr then

**Comment by zikalino** [comment]
@nitzmahone that's the first draft, i will still add idempotence soon.

**Comment by mattclay** [comment]
@zikalino You can run the sanity tests locally which will save you time waiting on Shippable to see if the tests pass. Let me know if you need assistance doing so.

**Comment by zikalino** [comment]
@mattclay thanks! yes, i know, i just didn't have proper env at that time.
i actually have one question. I am struggling with that import issue.
I mean everything runs correctly when using module and running integration tests.
But it keeps complaining about one import in sanity test environment...
Do you have any idea what could be a problem?

**Comment by mattclay** [comment]
@zikalino The import test requires that your Ansible module (and module_utils) can be imported and executed up until AnsibleModule instantiation without any dependencies outside the Python standard library available. That means all other imports must be in a `try/except ImportError` block. It also means that anything that depends on those imports must also occur after AnsibleModule instantiation, or a stub needs to be provided on import failure.

There are two main reasons for this:

1) Make sure that a module can provide a user friendly error message when dependencies are missing, instead of raising an import error which would trigger a rather generic module failure that is likely to be unhelpful to the user.
2) Make sure that the module validator is able to execute a module up through AnsibleModule instantiation so that module args can be validated.

The specific failure I'm seeing is for `AzureConfiguration`, which is then used by `class GenericRestClientConfiguration(AzureConfiguration)` when `AzureConfiguration` hasn't yet been defined (if the requirement is missing). You may be able to work around this by defining `AzureConfiguration = object` in the except block.

**Comment by zikalino** [comment]
@mattclay thank you very much for this info! This is really helpful! :-)

**Comment by zikalino** [comment]
@nitzmahone could you take a look?
I have added option for idempotence, by default idempotence is disabled.
I still plan to fix a few small details, but i think modules are in almost final shape, and I would appreciate any input so i can fix anything else needed by Monday.

**Comment by zikalino** [comment]
@nitzmahone just ping :-)

**Comment by mattclay** [comment]
@zikalino It looks like merge conflict resolution introduced unrelated changes to ignore.txt. There are many lines removed that are not Azure related.

**Comment by zikalino** [comment]
@mattclay yes, just noticed, i just wanted to revert ignore.txt changes... will fix it in a moment

**Comment by mattclay** [comment]
CI failure in integration test `test/integration/targets/azure_rm_resource/tasks/main.yml:57`:

> azure: azure_rm_resource : assert that=not output.changed

https://app.shippable.com/github/ansible/ansible/runs/66093/63/tests

**Review by nitzmahone — APPROVED** [review]
